libV2/tasks/Reindex.js

```diff
@@ -0,0 +1,177 @@
+        let lastMaster = null;
+        let lastMasterSize = null;
+
+        for await (const obj of metadata.listObjects(bucket)) {
+            if (obj.value.isDeleteMarker) {
+                // eslint-disable-next-line no-continue
+                continue;
+            }
+            count += 1;
+            size += obj.value['content-length'];
+
+            // If versioned, subtract the size of the master to avoid double counting
+            if (lastMaster && obj.name === lastMaster) {
+                logger.debug('Detected versioned key, subtracting master size', { lastMasterSize, key: obj.name });
+                size -= lastMasterSize;
+                count -= 1;
+                lastMaster = null;
+                lastMasterSize = null;
+            // Only save master versions
+            } else if (!obj.version) {
+                lastMaster = obj.name;
+                lastMasterSize = obj.value['content-length'];
+            }
+        }
+
+        return { size, count };
+    }
+
+    static async _indexMpuBucket(bucket) {
+        if (await metadata.bucketExists(bucket)) {
+            return ReindexTask._indexBucket(bucket);
+        }
+
+        return { size: 0, count: 0 };
+    }
+
+    async _fetchCurrentMetrics(level, resource) {
+        const timestamp = convertTimestamp(new Date().getTime());
+        const options = {
+            params: {
+                start: timestamp,
+                end: timestamp,
+                node: this._nodeId,
+                labels: {
+                    [level]: resource,
+                },
+            },
+            macro: 'utapi/getMetrics',
+        };
+        const res = await this._warp10.exec(options);
+        return { timestamp, value: JSON.parse(res.result[0]) };
+    }
+
+    async _updateMetric(level, resource, total) {
+        const { timestamp, value } = await this._fetchCurrentMetrics(level, resource);
+
+        const objectDelta = total.count - value.numberOfObjects[0];
+        const sizeDelta = total.size - value.storageUtilized[0];
+
+        if (objectDelta !== 0 || sizeDelta !== 0) {
+            logger.info('discrepancy detected in metrics, writing corrective record',
+                { [level]: resource, objectDelta, sizeDelta });
+
+            const record = new UtapiRecord({
+                objectDelta,
+                sizeDelta,
+                timestamp,
+            });
+
+            await this._warp10.ingest(
+                {
+                    className: 'utapi.repair.reindex',
+                    labels: {
+                        [level]: resource,
+                    },
+                    valueType: warp10RecordType,
+                },
+                [record],
+            );
+        }
+    }
+
+    async _execute() {
+        logger.debug('reindexing objects');
+
+        const accountTotals = {};
+        const ignoredAccounts = new Set();
+
+        await async.eachLimit(metadata.listBuckets(), 5, async bucket => {
```

Comment:
I'm surprised the `async` module functions can take real async functions returning promises and can be themselves async functions, but I'm not fully aware so it is likely correct!

Reply:
Me too! They were added in 3.x, I believe we mostly use 2.x in other projects

Reply:
OK, that's what I thought too :)